apps/client/src/pages/WrapperPage.js

```diff
@@ -17,14 +18,45 @@ import withChecker from "../hoc/withChecker";
 const WrapperPage = ({ checker, topic, resetChecker }) => {
   const sessionContext = useContext(SessionContext);
   const { slug, sttrFile } = topic;
-  const { finishedLocation, finishedQuestions } = sessionContext[slug] || false;
+  const { activeComponents, finishedComponents } = sessionContext[slug] || []; // At startup we don't have activeComponent or finishedComponents, so start with empty array.
+
+  // Only one component can be active at the same time.
+  const setActiveState = (component) => {
+    sessionContext.setSessionData([slug, { activeComponents: [component] }]);
+  };
+
+  // Pass a component name, or a array of components, this array or component will all be finished or removed.
+  const setFinishedState = (component, value) => {
+    let newFinishedComponents = finishedComponents || []; // If there are no finishedComponents start with a empty array.
+    const allComponents = Array.isArray(component) ? component : [component];
+    if (typeof value === "boolean" && value === false) {
+      newFinishedComponents =
+        finishedComponents?.filter((c) => !allComponents.includes(c)) || [];
+    } else {
+      newFinishedComponents = [...newFinishedComponents, ...allComponents];
+    }
+    // Save the new sessionData to the contex
+    sessionContext.setSessionData([
+      slug,
+      { finishedComponents: newFinishedComponents },
+    ]);
+  };
+
+  // @TODO: maybe this code can be refactored to one function, but i think this is really declarative. Lets discuss
+  const isActive = (component) => {
+    const allComponents = Array.isArray(component) ? component : [component];
+    return activeComponents?.includes(...allComponents) || false;
+  };
```

Comment:
If it's basically the same code than we can better use just one, or maybe something like this: ` const isFinished = (value) => isActive(value)`